### ISSUE TITLE
RFE: Bootstrap Loader

### ISSUE THREAD

**Issue by DALDEI** [uchuhimo/konf#3]
I would like to implement a lighter/better/simpler version of Spring cloud-config.
The concept is you have a 'bootstrap' loader which configures the source of further configurations.
This bootstrap load then could specify a URL or other source in which to read the rest of the configurations.

This can be done currently in 2 steps,  first create a Config for the bootstrap which would  have the a configuration specifying the Source/Loader to load.   Then parse that and create a new Config using your ConfigSpec  and <insert magic> add the   .withSourceFrom.xxxx  statements from the bootstrap process.
This was the rational for the previous pull request, so a URL could include the format making the bootstrap able to use a config file in any format easily.

Once a bootstrap is working then I am working on mini-server that will server config requests ... that has a bit more complexity in it as it needs to server to many clients and to pull out of the config files sub-trees of configurations.  Not sure how I plan on doing that yet as the ConfigSpec would not be in the server side only the loader, but its useful for it to be able to parse the files fully so it can pull out subtrees ... which requires a kind of ConfigSpec for that (or a bunch more low level code).

Ultimately Im aiming at kotlin-script based DSL that could be dynamically loaded.

But first step, a consistent config format to specify config sources and/or loaders.
Any suggestions welcome.

I have also implemented a basic s3 Loader essentially this:
```
fun getS3ObjectAsStream( s3url: String , region: String ) = ... 

...

fun Loader.s3(s3url: String , region: String): Config =
    config.withSource(provider.fromInputStream(getS3ObjectAsStream(s3url,region=region)))

...

val configSpec =  Config { addSpec(MyConfig) }
    .withSourceFrom.properties.s3("s3://mys3bucket/test/config.properties" , region="us-east-1" )

```

The goal being to put protected config files in S3 and then use the running app's authentication to load from them.   Ultimately the above would be configured in the bootstrap, possibly augmented with system properties.   In this case Im using it from within a AWS Lambda function which has limited access to the environment so it must get everything from outside or from env variables. 

I am working on enhancing the URL provider to accept the s3 scheme,  doing it the 'normal' way of adding new protocols to Java doesn't work in Lambda due to the security manager so it need be done with custom code.   (aka the getS3ObjectAsStream()) 

Also, thank you !  I have been looking for finding or writing a good config system and have had many dead-ends that were just not quite right -- (including my own) -- and finally found konf !! This is by far the cleanest and most generally useful config library Ive found yet so am basing my future apps on it.

**Comment by uchuhimo** [comment]
This feature can be implemented by the following steps (extract from your description):
1. Add a `BootstrapConfigSpec`, which defines the syntax of the bootstrapping process description
2. Use `Config { addSpec(BootstrapConfigSpec) }.withSourceFrom.file(...)` (or simply `BootstrapConfig.withSourceFrom.file(...)` if `BootstrapConfig` is defined in konf) to define the bootstrapping process
3. Implement and use `bootstrapConfig.bootstrap()` to get the config
To support this feature, `Item` should be extended to receive type parameter from outside, which is a small change. If we are on the same page, I can provide a prototype to you for further discussion when I have time off.

**Comment by DALDEI** [comment]
Thank you,  in parallel I implemented something similar in #5  .  I do not understand what you mean by Item extended to receive type parameter from outside. 
Could you explain also what you mean by 

`Implement and use bootstrapConfig.bootstrap() to get the config`

No hurry on any of this, I just wanted to document how I am using (or trying to use) this library.

**Comment by uchuhimo** [comment]
I mean to implement a function (suppose to name it as `bootstrap`), which can parse the instance of `BootstrapConfig` and use the description in `BootstrapConfig` to construct and return the desired config.
